exercises/react/example.cr

```diff
@@ -0,0 +1,76 @@
+class Cell
+  @dependents = [] of ComputeCell
+
+  def initialize(initial : Int32)
+    @value = initial
+  end
+
+  getter :value
+
+  protected getter :dependents
+end
+
+class InputCell < Cell
+  def value=(new_value : Int32)
+    @value = new_value
+    @dependents.each(&.propagate)
+    @dependents.each(&.fire_callbacks)
+  end
+end
+
+class ComputeCell < Cell
+  alias CallbackID = Int32
+
+  @callbacks_given = 0
+  @callbacks = {} of CallbackID => Int32 ->
+  @last_callback_value : Int32
+
+  def initialize(new_value : -> Int32, *inputs : Cell)
+    super(new_value.call)
+    @new_value = new_value
+    @last_callback_value = @value
+    inputs.each { |i| i.dependents << self }
+  end
+
+  def add_callback(&block : Int32 ->) : CallbackID
+    @callbacks_given += 1
+    @callbacks[@callbacks_given] = block
+    @callbacks_given
+  end
+
+  def remove_callback(id : CallbackID)
+    @callbacks.delete(id)
+  end
+
+  def propagate
+    v = @new_value.call
+    return if v == @value
+    @value = v
+    @dependents.each(&.propagate)
+  end
+
+  def fire_callbacks
+    return if @value == @last_callback_value
+    @last_callback_value = @value
+    @callbacks.each_value { |c| c.call(@value) }
+    @dependents.each(&.fire_callbacks)
+  end
+end
+
+class Reactor
+  # Note that the Reactor does not keep any extra state *in this example solution*.
+  # In other solutions, it's possible that Reactors may keep track of their cells,
+  # take reponsibility for value propagation and callback firing, etc.
+
+  def create_input(initial : Int32) : InputCell
+    InputCell.new(initial)
+  end
+
+  def create_compute(input : Cell, &block : Int32 -> Int32) : ComputeCell
+    ComputeCell.new(->{ block.call(input.value) }, input)
+  end
+
+  def create_compute(input1 : Cell, input2 : Cell, &block : (Int32, Int32) -> Int32) : ComputeCell
```

Comment:
I wished that I could create a general function of the form `create_compute(*inputs : Cell, &block : Something -> Int32) : ComputeCell`.

However, I was not able to do such a thing. I think the general problem is you cannot write a type for "A function that takes as many inputs as there are values in `inputs`.

Now, you could do `def create_compute(*inputs : Cell, &block : Array(Int32) -> Int32) : ComputeCell` (and the below line would be something like `ComputeCell.new(inputs.map(&.value).to_a}, input1, input2)`. However, now note that we have lost the ability to check that the types match. Someone could create a compute cell with 1 input but whose compute function takes 2 inputs, or vice versa.

So I did it this way to maintain that the compiler can check that the arities match up. Obviously the disadvantage is if we want compute cells that take 3 inputs we have to write a third function, etc.